lib/galaxy/datatypes/data.py

```diff
@@ -520,8 +520,12 @@ def convert_dataset(self, trans, original_dataset, target_type, return_output=Fa
                 params[value.name] = deps[value.name]
             elif value.type == 'data':
                 input_name = key
-
+        if target_context:
+            for key, value in target_context.items():
```

Comment:
Can you add a comment here about this - I'd be very confused by this section otherwise.

Also - is it better to copy this or give it an empty/default state then a copy of the parent one? I don't think the conversion of a dataset for assembly say should be scheduled in the same way as an assembly job.